File: src/installer/java/org/eclim/plugin/pydev/Application.java
/**
 * Copyright (C) 2005 - 2008  Eric Van Dewoestine
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.eclim.plugin.pydev;

import org.eclipse.core.runtime.IPlatformRunnable;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.NullProgressMonitor;

import org.eclipse.swt.widgets.Display;

import org.python.pydev.core.IInterpreterManager;

import org.python.pydev.plugin.PydevPlugin;

/**
 * Entry point for installer application.
 *
 * @author Eric Van Dewoestine
 * @version $Revision$
 */
public class Application
  implements IPlatformRunnable
{
  private static final String LIST = "list";
  private static final String SET = "set";

  /**
   * Runs this runnable with the supplied args.
   *
   * @param _args The arguments (typically a String[]).
   * @return The result (returned Integer is treated as exit code).
   */
  public Object run (Object _args)
    throws Exception
  {
    try{
      // create the eclipse workbench.
      org.eclipse.ui.PlatformUI.createAndRunWorkbench(
          Display.getDefault(), new WorkbenchAdvisor());

      String[] params = (String[])_args;

      IInterpreterManager manager = PydevPlugin.getPythonInterpreterManager();
      if(LIST.equals(params[params.length - 1])){
        String [] executables = manager.getInterpretersFromPersistedString(
            manager.getPersistedString());
        for (int ii = 0; ii < executables.length; ii++){
          System.out.println(executables[ii]);
        }
      }else if(SET.equals(params[params.length - 2])){
        String interpreter = params[params.length - 1];
        manager.getInterpreterInfo(interpreter, new NullProgressMonitor());
        manager.setPersistedString(
            manager.getStringToPersist(new String[]{interpreter}));
        PydevPlugin.getDefault().savePluginPreferences();
      }
    }catch(Throwable t){
      return new Integer(1);
    }

    return new Integer(0);
  }

  private static class WorkbenchAdvisor
    extends org.eclipse.ui.application.WorkbenchAdvisor
  {
    /**
     * {@inheritDoc}
     */
    public String getInitialWindowPerspectiveId ()
    {
      return null;
    }

    /**
     * {@inheritDoc}
     */
    public boolean openWindows ()
    {
      return false;
    }
  }
}

File: doc/extension/eclim/sphinx/application.py
"""
Copyright (C) 2005 - 2010  Eric Van Dewoestine

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
from os import path
import sys

from sphinx.application import Sphinx as SphinxBase, ENV_PICKLE_FILENAME
from sphinx.util.console import bold
from sphinx.util.osutil import ENOENT

from eclim.sphinx.environment import EclimBuildEnvironment, EclimHtmlBuildEnvironment

class Sphinx(SphinxBase):
  ENV_CLASSES = {
    'eclim': EclimHtmlBuildEnvironment,
    'vimdoc': EclimBuildEnvironment,
  }

  def __init__(
    self, srcdir, confdir, outdir, doctreedir, buildername,
    confoverrides=None, status=sys.stdout, warning=sys.stderr,
    freshenv=False, warningiserror=False, tags=None
  ):
    self.buildername = buildername
    super(Sphinx, self).__init__(
      srcdir, confdir, outdir, doctreedir, buildername, confoverrides,
      status, warning, freshenv, warningiserror, tags)

  def _init_env(self, freshenv):
    """
    Copied from sphinx.application.Sphinx with BuildEnvironment replaced with
    the appropriate eclim build environment class.
    """
    env_class = self.ENV_CLASSES[self.buildername]

    if freshenv:
      #self.env = BuildEnvironment(self.srcdir, self.doctreedir, self.config)
      self.env = env_class(self.srcdir, self.doctreedir, self.config)
      self.env.find_files(self.config)
      for domain in self.domains.keys():
        self.env.domains[domain] = self.domains[domain](self.env)
    else:
      try:
        self.info(bold('loading pickled environment... '), nonl=True)
        #self.env = BuildEnvironment.frompickle(self.config,
        #    path.join(self.doctreedir, ENV_PICKLE_FILENAME))
        self.env = env_class.frompickle(self.config,
            path.join(self.doctreedir, ENV_PICKLE_FILENAME))
        self.env.domains = {}
        for domain in self.domains.keys():
          # this can raise if the data version doesn't fit
          self.env.domains[domain] = self.domains[domain](self.env)
        self.info('done')
      except Exception, err:
        if type(err) is IOError and err.errno == ENOENT:
          self.info('not yet created')
        else:
          self.info('failed: %s' % err)
        return self._init_env(freshenv=True)

    self.env.set_warnfunc(self.warn)
